src/cockpit/peer.py

```diff
@@ -142,3 +142,41 @@ def do_channel_control(self, channel: str, command: str, message: Dict[str, obje
 
     def do_channel_data(self, channel: str, data: bytes) -> None:
         self.write_channel_data(channel, data)
+
+
+class PeerRoutingRule(RoutingRule, PeerStateListener):
+    predicate: Callable[[Dict[str, object]], bool]
+    args: List[str]
+    env: Dict[str, str]
+    peer: Optional[Peer]
+
+    def __init__(self, router: Router, args: List[str], env: Dict[str, str], predicate: Callable[[Dict[str, object]], bool]):
+        super().__init__(router)
+
+        self.args = args
+        self.env = env
+        self.predicate = predicate
+        self.peer = None
+
+    def apply_rule(self, options: Dict[str, object]) -> Optional[Peer]:
+        if not self.predicate(options):
+            return None
+
+        if self.peer is None:
+            try:
+                peer = Peer(self.router, self.args[0], self)
+                peer.spawn(self.args, self.env)
+                peer.freeze_endpoint()
+                self.peer = peer
+            except OSError:
+                raise RoutingError('blah-blah')
+
+        return self.peer
+
+    def peer_state_changed(self, peer: Peer, event: str, exc: Optional[Exception] = None):
```

Comment:
Note to self: This isn't called from anywhere in this commit; This is called from `Peer`.